### PR TITLE
Allow event monitoring to ignore failing event accessors

### LINKED ISSUE #1948: Exceptions in event accessors fail tests when using .Monitor()
### Description

When an event handler accessor has an explicit implementation, which throws an exception on the setter, `.Monitor()` is failing the test completely.

**What is my motivation?**

I need to test the class behavior without changing the other base event handler accessors behavior. It could cause unwanted side effects. But I do not need to test this event. I just wanted to test PropertyChanged events being raised. Therefore, I would like to have the benefit of having the full class available to write expressions in the assertion.

**Suggestion:**

Make some more configuration options for the Monitor. Like ignoring failing events, or just subscribing to a subset of interfaces with those events, or putting some events on ignore, without changing the generic Type, which is used to assert / monitor.

**Benefit:**

Using the great syntax you introduced in monitor on brown field projects or if you are not able to change the behavior of some event accessors. Therefore, write better and cleaner test code from the beginning.

### Complete minimal example reproducing the issue

```csharp
using System.ComponentModel;
using Xunit;

namespace FluentAssertions.Tests;

class FailingEventHandlerSubscription : INotifyPropertyChanged
{
    private event PropertyChangedEventHandler? InnerPropertyChanged;
    public event PropertyChangedEventHandler? PropertyChanged
    {
        add { throw new System.Exception("This is sadly happening due to anonymous method"); }
        remove { InnerPropertyChanged -= value; }
    }

    private object prop;

    public object Property
    {
        get => prop;
        set
        {
            prop = value;
            this.InnerPropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Property)));
        }
    }
}

public class UnitTest2
{
    [Fact]
    public void TestEventHandler()
    {
        // Arrange
        var cut = new FailingEventHandlerSubscription();

       // will fail, because the exception is thrown, when the EventHandler class is subscribing the event.
        using var monitored = cut.Monitor();

        // Act
        cut.Property = new object();

        // Assert
        // If I just use the INotifyPropertyChanged interface, I cannot write this awesome expression.
        monitored.Should().RaisePropertyChangeFor(m => m.Property);
    }
}
```

### Expected behavior:

FluentAssertions is able to have options in the monitor to ignore failing event handlers.

### Actual behavior:

It throws an exception when .Monitor() is called and I can do nothing about it, event though I do not want to test this specific event.

### Versions

FluentAssertions Version 6.7.0

### Additional Information

I already implemented my custom implementation of the monitor and would offer to contribute my suggestion above to the FluentAssertions project. I just wanted to talk to you first, if you would approve of such an enhancement of the monitor or if you disagree on this feature for some reason.

So let me know if you agree or disagree.

cheers!

### PR DESCRIPTION
**Pull request by ITaluone**
This finishes #1954 

Closes #1948 

**Note:** All credits goes to @apazureck!

## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).
    * [x] Please also run `./build.sh --target spellcheck` or `.\build.ps1 --target spellcheck` before pushing and check the good outcome

### CODE REVIEW

File: Src/FluentAssertions/Events/EventMonitorOptions.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    /// <summary>
+    /// Func used to generate the timestamp.
+    /// </summary>
+    internal Func<DateTime> TimestampProvider { get; private set; } = () => DateTime.UtcNow;
+
+    /// <summary>
+    /// When called it will ignore event accessor Exceptions.
+    /// </summary>
+    /// <returns>The options instance for method stacking.</returns>
+    public EventMonitorOptions IgnoreEventAccessorExceptions()
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Events/EventMonitorOptions.cs:29]: 🔧 Similar to `SelfReferenceEquivalencyOptions`, we should `IgnoringEventAccessorExceptions`. Same for `RecordEventsWithBrokenAccessor`

File: Src/FluentAssertions/Events/EventMonitorOptions.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        ShouldRecordEventsWithBrokenAccessor = true;
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the timestamp provider. By default it is <see cref="DateTime.UtcNow"/>.
+    /// </summary>
+    /// <param name="timestampProvider">The timestamp provider.</param>
+    /// <returns>The options instance for method stacking.</returns>
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Events/EventMonitorOptions.cs:49]: 🔧 No need to explicitly document the return type of a fluent API

File: docs/_pages/releases.md [hunk 3]
```diff
@@ -16,6 +16,8 @@ sidebar:
 * Add support for NUnit 4 - [#2483](https://github.com/fluentassertions/fluentassertions/pull/2483)
 * Added `NotBeIn` to check if a `DateTime` is not in a given `DateTimeKind` - [#2536](https://github.com/fluentassertions/fluentassertions/pull/2536)
 * Added `BeNaN` and `NotBeNaN` for assertions on `float` and `double` - [#2606](https://github.com/fluentassertions/fluentassertions/pull/2606)
+* Added options to `.Monitor(...)` extension - [#1954](https://github.com/fluentassertions/fluentassertions/pull/1954)
```
Comment (dennisdoomen) [inline docs/_pages/releases.md:19]: 🔧 We can remove this one as it has no functional purpose. It's only there to support the `IgnoreEventAccessorExceptions` and `RecordEventsWithBrokenAccessor` methods

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Thanks for taking over the work. Just some minor comments to address.

**Review by dennisdoomen — APPROVED** [review]

File: Src/FluentAssertions/AssertionExtensions.cs [hunk 4]
```diff
@@ -873,14 +873,29 @@ public static TaskCompletionSourceAssertions<T> Should<T>(this TaskCompletionSou
...
+    }
+
+    /// <summary>
+    /// Starts monitoring <paramref name="eventSource"/> for its events using the given <paramref name="configureOptions"/>.
+    /// </summary>
+    /// <param name="eventSource">The object for which to monitor the events.</param>
+    /// <param name="configureOptions">
+    /// Options to configure the EventMonitor.
+    /// </param>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="eventSource"/> is Null.</exception>
```
Comment (jnyrup) [inline Src/FluentAssertions/AssertionExtensions.cs:891]:
```suggestion
    /// <exception cref="ArgumentNullException"><paramref name="eventSource"/> is <see langword="null"/>.</exception>
    /// <exception cref="ArgumentNullException"><paramref name="configureOptions"/> is <see langword="null"/>.</exception>
```

File: docs/_pages/releases.md [hunk 5]
```diff
@@ -74,6 +75,7 @@ sidebar:
   * `IEquivalencyAssertionOptions` to `IEquivalencyOptions`
   * `SelfReferenceEquivalencyAssertionOptions<TSelf>` to `SelfReferenceEquivalencyOptions<TSelf>`
 * Allow fluently calling `WithoutMatchingRules` and `WithoutSelectionRules` while using `BeEquivalentTo` - [#2457](https://github.com/fluentassertions/fluentassertions/pull/2457)
+* Removed `utcNow` overload for `.Monitor<T>()` - [#1954](https://github.com/fluentassertions/fluentassertions/pull/1954)
```
Comment (jnyrup) [inline docs/_pages/releases.md:78]:
```suggestion
* Removed `utcNow` overload for `.Monitor<T>()` - [#2629](https://github.com/fluentassertions/fluentassertions/pull/2629)
```

Comment (ITaluone) [reply docs/_pages/releases.md:78]: Ou.. thanks for head up 👍

**Review by jnyrup — APPROVED** [review]

**Comment by ITaluone** [comment]
wait a sec

**Comment by ITaluone** [comment]
had a litte incident with rebase/force pushing 🤦‍♂️

**Pull request merged**